Hash Table 0.1

MShepelin opened the pull request:
All tests passed
https://official.contest.yandex.ru/contest/17022/run-report/30816151/

MShepelin committed:
Hash Table 0.1

All tests passed

MShepelin committed:
names and operator [] fixed

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,192 @@
+    table_type table_;
+
+    const size_type kMinSize = 16;
+    const double alpha_ = 2;
```
Why double. And kAlpha.

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,192 @@
+
+    void rehash() {
+        size_ *= 2;
+        table_type temp_table(size_, { list_of_pairs_.end() });
```
```c++
table_.clear();
table_resize(size_);
```
Should works without temp_table.

Also temp table miss naming rules: https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,192 @@
+        size_ = kMinSize;
+        elements_stored_ = 0;
+        list_of_pairs_= list_type(0);
+        table_ = table_type(size_, { list_of_pairs_.end() });
```
Move to initialization list.

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,192 @@
+        }
+    }
+
+    HashMap(std::initializer_list<std::pair<KeyType, ValueType>> init_list, Hash hasher = Hash()) : HashMap(hasher) {
```
init_list - naming rules
Delegate to previous constructor.

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    void erase(KeyType key) {
+        iterator list_iterator = find(key);
```
list_iterator - meaningless name

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,192 @@
+    const size_type kMinSize = 16;
+    const double alpha_ = 2;
+    size_type elements_stored_;
+    size_type size_;
```
size_ is not a size actually. Please give more meaningful name.

MShepelin committed:
Decoration problems solved

Checked with cpplint.py: https://raw.githubusercontent.com/google/styleguide/gh-pages/cpplint/cpplint.py
Names fixed for more meaningful, number of spaces for tabs changed to 2

MShepelin committed:
temp_table name fixed

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,197 @@
+
+  void rehash() {
+    slots_available_ *= 2;
+    table_type table_copy(slots_available_, { list_of_pairs_.end() });
```
With this table, you will copy all iterators twice. Once on rehash and once, on `table_ = table_copy`, please copy only once. 

`table_.clear();` will help you.

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,197 @@
+  const_iterator find(KeyType key) const {
+    size_type key_hash = hasher_(key);
+    for (auto list_iterator : table_[key_hash % slots_available_]) {
+      if (list_iterator != list_of_pairs_.end() &&
```
Why do you need `list_of_pairs_.end()` in table_?

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,197 @@
+
+    if (elements_stored_ * kAlpha > slots_available_) {
+      rehash();
+    }
```
Copy&paste from insert.

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,197 @@
+    return list_of_pairs_.end();
+  }
+
+  void insert(std::pair<KeyType, ValueType> key_and_value) {
```
Please read requirement in the chat.

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,197 @@
+    slots_available_ = kMinSize;
+    elements_stored_ = 0;
+    list_of_pairs_ = list_type(0);
+    table_ = table_type(slots_available_, { list_of_pairs_.end() });
```
`list_of_pairs_.end()` is useless here.

MShepelin committed:
Mistakes fixed

rehash() worked to long because of extra copying
insert() and operator[] had copied code

feldsherov commented on hash_map.h:
```diff
@@ -9,189 +9,175 @@ template<
+      slots_available_ *= 2;
+      table_ = table_type(slots_available_);
+
+      for (iterator iterator_to_key_and_value = list_of_pairs_.begin();
```
range based for

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,183 @@
+      size_type key_hash = hasher_(key);
+
+      auto& list_of_iters = table_[key_hash % slots_available_];
+      for (auto table_list_it = list_of_iters.begin();
```
может тоже range based for?

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,183 @@
+    template<class ForwardIt>
+    HashMap(ForwardIt first, ForwardIt last, Hash hasher = Hash()) :
+      HashMap(hasher) {
+      while (first != last) {
```
Можно вот это заиспользовать: https://ru.cppreference.com/w/cpp/algorithm/for_each

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,183 @@
+      return list_of_pairs_.end();
+    }
+
+    const_iterator find(KeyType key) const {
```
Алиса умеет убирать копипасту, спроси у нее.

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,183 @@
+    const_iterator find(KeyType key) const {
+      size_type key_hash = hasher_(key);
+      for (auto list_iterator : table_[key_hash % slots_available_]) {
+        if (list_iterator != list_of_pairs_.end() &&
```
давай вместо `list_of_pairs_.end()` вызывать `end()`?
Везде в коде, где можем.

feldsherov commented on hash_map.h:
```diff
@@ -0,0 +1,183 @@
+
+    const ValueType& at(KeyType key) const {
+      const_iterator list_iterator = find(key);
+      if (list_iterator == list_of_pairs_.end())
```
скобочки

MShepelin committed:
Problem fixed

Tests check: https://official.contest.yandex.ru/contest/17022/run-report/31367830/

feldsherov reviewed (approved):
ок
